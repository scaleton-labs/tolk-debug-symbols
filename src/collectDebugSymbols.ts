import { TolkCompilerConfig } from '@ton/tolk-js';
import { DebugSymbols } from './types';
import { extractSymbolsFromSourceCode } from './extractSymbolsFromSourceCode';
import { extractSymbolsFromCompiler } from './extractSymbolsFromCompiler';

export async function collectDebugSymbols(
  config: TolkCompilerConfig,
): Promise<DebugSymbols> {
  const [symbolsFromCompiler, snapshot] =
    await extractSymbolsFromCompiler(config);

  const symbolsFromSourceCode = await extractSymbolsFromSourceCode(snapshot);

  return {
    ...symbolsFromCompiler,
    ...symbolsFromSourceCode,
  };
}
