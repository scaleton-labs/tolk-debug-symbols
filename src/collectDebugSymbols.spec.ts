import { TolkCompilerConfig } from '@ton/tolk-js';
import { collectDebugSymbols } from './collectDebugSymbols';

describe('collectDebugSymbols', () => {
  it('collects information about procedures and globals', async () => {
    const config: TolkCompilerConfig = {
      entrypointFileName: 'main.tolk',
      fsReadCallback: () => `
        global a: int;
        global b: cell;
        global c: slice;

        @inline
        fun throwInline() {
          throw 1;
        }

        @inline_ref
        fun throwInlineRef() {
          throw 2;
        }

        @method_id(88)
        fun throwGet() {
          throw 3;
        }

        fun onInternalMessage() {
          throw 4;
        }
      `,
    };

    const debugSymbols = await collectDebugSymbols(config);
    expect(debugSymbols).toMatchSnapshot();
  });

  it('collects information about constants', async () => {
    const config: TolkCompilerConfig = {
      entrypointFileName: 'main.fc',
      fsReadCallback: () => `
        const h: int = 0xBEEF;
        const k: int = 1000;
        const l: slice = "hello";
        const m: slice = "cafe"s;
        const n: slice = "Ef8zMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzM0vF"a;
        const o: int = "ping"u;
        const p: int = "tranfser"h;
        const q: int = "tranfser"H;
        const r: int = "tranfser"c;

        // expressions aren't supported
        const i: int = -0xCAFE;
        const j: int = -1000;
        const s: int = k * j;

        get getConstants() {
          return (h, i, j, k, l, m, n, o, p, q, r);
        }

        fun onInternalMessage(inMsgBody: slice) {
          return ();
        }
      `,
    };

    const debugSymbols = await collectDebugSymbols(config);
    expect(debugSymbols).toMatchSnapshot();
  });

  it('works with empty functions', async () => {
    const config: TolkCompilerConfig = {
      entrypointFileName: 'main.fc',
      fsReadCallback: () => `
        fun main() {}
      `,
    };

    const debugSymbols = await collectDebugSymbols(config);
    expect(debugSymbols).toMatchSnapshot();
  });

  it('works with untyped constants', async () => {
    const config: TolkCompilerConfig = {
      entrypointFileName: 'main.fc',
      fsReadCallback: () => `
        const a = 1000;
        fun main() {}
      `,
    };

    const debugSymbols = await collectDebugSymbols(config);
    expect(debugSymbols).toMatchSnapshot();
  });
});
