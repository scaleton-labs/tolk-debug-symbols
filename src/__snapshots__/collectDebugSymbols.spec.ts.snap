// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`collectDebugSymbols collects information about constants 1`] = `
{
  "constants": [
    {
      "declaration": "0xBEEF",
      "name": "h",
      "type": "int",
      "value": "48879",
    },
    {
      "declaration": "1000",
      "name": "k",
      "type": "int",
      "value": "1000",
    },
    {
      "declaration": ""hello"",
      "name": "l",
      "type": "slice",
      "value": "x{68656c6c6f}",
    },
    {
      "declaration": ""cafe"s",
      "name": "m",
      "type": "slice",
      "value": "x{cafe}",
    },
    {
      "declaration": ""Ef8zMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzM0vF"a",
      "name": "n",
      "type": "slice",
      "value": "x{9fe6666666666666666666666666666666666666666666666666666666666666667_}",
    },
    {
      "declaration": ""ping"u",
      "name": "o",
      "type": "int",
      "value": "1885957735",
    },
    {
      "declaration": ""tranfser"h",
      "name": "p",
      "type": "int",
      "value": "1951942752",
    },
    {
      "declaration": ""tranfser"H",
      "name": "q",
      "type": "int",
      "value": "52624272503553231905816075160042524215763620945865102247584047589936797967283",
    },
    {
      "declaration": ""tranfser"c",
      "name": "r",
      "type": "int",
      "value": "1025402358",
    },
  ],
  "globals": [],
  "procedures": [
    {
      "cellHash": "a0eacc9676d4e13d5bade93200eee2734baf0ad2256df4ee1983c5e6ce29c388",
      "methodId": 0,
      "name": "main",
    },
    {
      "cellHash": "40956732df8dbebee9f697ef04dcfc498b671e7167ca2b898e85dbc737089012",
      "methodId": 78674,
      "name": "getConstants",
    },
  ],
}
`;

exports[`collectDebugSymbols collects information about procedures and globals 1`] = `
{
  "constants": [],
  "globals": [
    {
      "index": 1,
      "name": "a",
    },
    {
      "index": 2,
      "name": "b",
    },
    {
      "index": 3,
      "name": "c",
    },
  ],
  "procedures": [
    {
      "cellHash": "b1b3726fb943fec8e144bb0e470a74a891fa4948fbc50b85ca1981e5444dfb4b",
      "methodId": 0,
      "name": "main",
    },
    {
      "cellHash": "f8664b544b6b2451eb1826d6fc9c70fcd94f89a092391fa68355ea1c08b51c0d",
      "methodId": 1,
      "name": "throwInline",
    },
    {
      "cellHash": "7219c22ada21ad364e5e0844e953e759e275b68603c645c29c15ffa57e2d47e3",
      "methodId": 2,
      "name": "throwInlineRef",
    },
    {
      "cellHash": "588d70db264b8d53c52e563524749977fd62985d6bda14e3a9b83dd5ba9c9041",
      "methodId": 88,
      "name": "throwGet",
    },
  ],
}
`;
